### ISSUE TITLE
How to create custom rules to scan scerts in .yaml or .properties file 

### ISSUE THREAD

**Issue by pvimalmec** [zricethezav/gitleaks#1078]
**Describe the bug**
With below custom rule, we are trying to find the secerts in .yaml|.yml|.properties file, but its not working as expected.
[[rules]]
description = "Generic Secerts in resources"
regex = '''(?i)(?:key|api|token|secret|client|passwd|password|auth|access)(?:[0-9a-z\-_\t .]{0,20})(?:[\s|']|[\s|"]){0,3}(?:=|>|:=|\|\|:|<=|=>|:)(?:'|\"|\s|=|\x60){0,5}([0-9a-z\-_.=]{10,150})(?:['|\"|\n|\r|\s|\x60|;]|$)'''
path = '''(?i)(\.prop*|\.yml|\.yaml)'''
tags = ["key", "properties"]

**To Reproduce**

**Expected behavior**

**Screenshots**

**Basic Info (please complete the following information):**
 - OS:
 - Gitleaks Version: 8.15.2

**Additional context**

cc @zricethezav

**Comment by weineran** [comment]
> but its not working as expected.

Hi @pvimalmec . Can you describe the behavior you are seeing with your custom rule?

**Comment by pvimalmec** [comment]
> > but its not working as expected.
> 
> Hi @pvimalmec . Can you describe the behavior you are seeing with your custom rule?

@weineran when used the above custom rule, its scanning all the files within my project (like *.java, Jenkinsfile, *.json and so on). I would like to scan only the  .yaml or .properties file.

I have tried to separate the rules like below 
[[rules]]
description = "Generic Secerts in Properties file"
regex = '''(?i)(?:key|api|token|secret|client|passwd|password|auth|access)(?:[0-9a-z\-_\t .]{0,20})(?:[\s|']|[\s|"]){0,3}(?:=|>|:=|\|\|:|<=|=>|:)(?:'|\"|\s|=|\x60){0,5}([0-9a-z\-_.=]{10,150})(?:['|\"|\n|\r|\s|\x60|;]|$)'''
path = '''(?i)(\.prop*|\.yml|\.yaml)'''
tags = ["key", "properties"]

[[rules]]
description = "yaml secrets"
regex = '''(?i)(password|token|secret|client|passwd|enc.key|auth.pass):\s+(.*)'''
path = '''(?i)(\.yml|\.yaml)'''
tags = ["key", "yaml"]

When i tried this, its not finding the secerts in *. properties  files. Its give us zero leaks. but i have many secerts matching the pattern in my application.properties

**Comment by weineran** [comment]
Understood. Can you post the verbose logs from gitleaks when you run it with the above config?

**Comment by pvimalmec** [comment]
> Understood. Can you post the verbose logs from gitleaks when you run it with the above config?

Below is the verbose logs when i used two separate rules
`
    ○
    │╲
    │ ○
    ○ ░
    ░    gitleaks

9:54PM INF 400 commits scanned.
9:54PM INF scan completed in 1.54s
9:54PM INF no leaks found
`

**Comment by weineran** [comment]
Hmmm is there nothing in the log output that mentions the config file?

Can you post the command you are running?

**Comment by pvimalmec** [comment]
> [[rules]]
> description = "Generic Secerts in Properties file"
> regex = '''(?i)(?:key|api|token|secret|client|passwd|password|auth|access)(?:[0-9a-z\-_\t .]{0,20})(?:[\s|']|[\s|"]){0,3}(?:=|>|:=|\|\|:|<=|=>|:)(?:'|\"|\s|=|\x60){0,5}([0-9a-z\-_.=]{10,150})(?:['|\"|\n|\r|\s|\x60|;]|$)'''
> path = '''(?i)(\.prop*|\.yml|\.yaml)'''
> tags = ["key", "properties"]
> 
> [[rules]]
> description = "yaml secrets"
> regex = '''(?i)(password|token|secret|client|passwd|enc.key|auth.pass):\s+(.*)'''
> path = '''(?i)(\.yml|\.yaml)'''
> tags = ["key", "yaml"]



> Hmmm is there nothing in the log output that mentions the config file?
> 
> Can you post the command you are running?

`gitleaks detect  -c .\gitleaks.toml --verbose `

**Comment by zricethezav** [comment]
@pvimalmec add ` --log-level=debug` to see debug logs

**Comment by weineran** [comment]
Can you post an example of a secret you'd expect it to detect in a .properties or .yml file?

You mentioned that it's scanning other files such as .java, .json, etc. Can you post an example of that?

Or even better, if you could post an example repo that we can use to reproduce the issue, that might be best.

**Comment by pvimalmec** [comment]
Here is the repo which scans only the secerts in yaml file.
https://github.com/pvimalmec/gitleaks-repo.git

**Comment by weineran** [comment]
@pvimalmec As you can see in the logs you posted in the Readme, the scan successfully detected 3 leaks: https://github.com/pvimalmec/gitleaks-repo/blob/main/README.md?plain=1#L15

If you want to see details about the leaks, add the `-v, --verbose` flag.

Here is the full output on my system, running the latest version of gitleaks:
```
$ ../../gitleaks/gitleaks version
v8.15.2
$ ../../gitleaks/gitleaks detect -c gitleaks.toml --log-level=debug -v

    ○
    │╲
    │ ○
    ○ ░
    ░    gitleaks

9:05AM DBG using gitleaks config gitleaks.toml from `--config`
9:05AM DBG executing: /usr/bin/git -C . log -p -U0 --full-history --all
Finding:     password: test1234
Secret:      password: test1234
RuleID:
Entropy:     3.794653
File:        application.yaml
Line:        4
Commit:      b75ace2f377f372c3a40f93fd00c5829fae73f1a
Author:      periyasamy003
Email:       vimalraj.periyasamy2@tre.se
Date:        2023-01-20T09:59:16Z
Fingerprint: b75ace2f377f372c3a40f93fd00c5829fae73f1a:application.yaml::4

Finding:     password: test123
Secret:      password: test123
RuleID:
Entropy:     3.690116
File:        application.yaml
Line:        4
Commit:      45e9ffe9acc7135fcc08ee2d307b096df525d7bb
Author:      periyasamy003
Email:       vimalraj.periyasamy2@tre.se
Date:        2023-01-20T09:57:33Z
Fingerprint: 45e9ffe9acc7135fcc08ee2d307b096df525d7bb:application.yaml::4

Finding:     password: dbpwd
Secret:      password: dbpwd
RuleID:
Entropy:     3.189898
File:        application.yaml
Line:        10
Commit:      45e9ffe9acc7135fcc08ee2d307b096df525d7bb
Author:      periyasamy003
Email:       vimalraj.periyasamy2@tre.se
Date:        2023-01-20T09:57:33Z
Fingerprint: 45e9ffe9acc7135fcc08ee2d307b096df525d7bb:application.yaml::10

9:05AM INF 3 commits scanned.
9:05AM DBG Note: this number might be smaller than expected due to commits with no additions
9:05AM INF scan completed in 52.8ms
9:05AM WRN leaks found: 3
```

**Comment by pvimalmec** [comment]
@weineran  I agree but it actually show only the secerts from the yaml (application.yaml) not from the [application.properties][](https://github.com/pvimalmec/gitleaks-repo/blob/main/application.properties).

Hope you got what im looking for. gitleaks should scan the secret in application.properties https://github.com/pvimalmec/gitleaks-repo/blob/main/application.properties#L2

**Comment by weineran** [comment]
In order to debug your issue, I can suggest a couple next steps:
1. Make sure each rule in your config includes an `id`. The first two rules in your config do not have an `id`. As a result, the `RuleID` fields in the [findings above](https://github.com/zricethezav/gitleaks/issues/1078#issuecomment-1398604954) are blank. So it's unclear which rule is detecting those secrets. It looks like the secrets are being detected by the second rule in the config, but you'll want to confirm.
2. Ask yourself which rule you expect to detect the secrets in .properties file. Then use https://regex101.com/ to test out your rules and debug them until they detect the secrets that you want them to detect.

Another option is to just start with the [generic-api-key rule in the default config](https://github.com/zricethezav/gitleaks/blob/master/config/gitleaks.toml#L485), and tweak it until it works for you. For example, you could change `{10,150}` to `{5,150}`, and that should catch all the secrets in your example repo.

I hope that helps.

**Comment by pvimalmec** [comment]
@weineran  Thanks,  Is it not possible to scan only specfic files rather doing scan all files available in the repo/project?

**Comment by weineran** [comment]
It should be possible.

In your example repo, create a copy of your application.yaml file and rename it application.txt. You should see that the second rule in your config will detect secrets in the .yaml file but not in the .txt file.

Can you confirm?

**Comment by pvimalmec** [comment]
Yes, i can confirm it doesnt scan in .txt file?

**Comment by weineran** [comment]
Great, so that means it is behaving as desired, correct?

**Comment by weineran** [comment]
@pvimalmec Following up on this issue, are the custom rules working as desired or did I misunderstand?
